### PR TITLE
Ensure AccessionWF is kicked off once

### PR DESCRIPTION
**Pull request by edsu**
## Why was this change made? 🤔

We had a situation where large file deposits were triggering a race
condition where multiple accession workflows resulted in techmd getting
confused about missing files.

See https://github.com/sul-dlss/happy-heron/issues/3007

This change ensures that only one AccessionWF is started when staging
from ActiveStorage or Globus. The code assumes that a deposit
stages from ActiveStorage or Globus, but not both.

## How was this change tested? 🤨

Integration tests in stage.

⚡ ⚠ If this change has cross service impact, including writes to shared file systems, ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** (e.g. create_object_h2_spec) and/or test in [stage|qa] environment, in addition to specs. ⚡

### CODE REVIEW

File: app/jobs/ingest_job.rb [hunk 1]
```diff
@@ -48,12 +48,16 @@ def perform(model_params:, background_job_result:, signed_ids: {}, globus_ids: {
...
+    if signed_ids.any?
+      StageBlobs.stage(signed_ids, druid) do
+        Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority) if start_workflow
+      end
+    elsif globus_ids.any?
+      StageGlobus.stage(globus_ids, druid) do
+        Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority) if start_workflow
+      end
+    elsif start_workflow
+      Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority)
```
Comment (edsu) [inline app/jobs/ingest_job.rb:60]: I wasn't sure if this was needed. Do we ever want to kick off accessioning when there are no files to stage?

File: app/jobs/update_job.rb [hunk 2]
```diff
@@ -57,11 +58,15 @@ def perform(model_params:,
...
+    if signed_ids.any?
+      StageBlobs.stage(signed_ids, model.externalIdentifier) do
+        version_or_accession(object_client, model, existing, versioning_params)
+      end
+    elsif globus_ids.any?
+      StageGlobus.stage(globus_ids, model.externalIdentifier) do
+        version_or_accession(object_client, model, existing, versioning_params)
+      end
+    else
       version_or_accession(object_client, model, existing, versioning_params)
```
Comment (edsu) [inline app/jobs/update_job.rb:70]: It seems like this is needed when updating metadata only?

File: app/jobs/ingest_job.rb [hunk 3]
```diff
@@ -48,12 +48,16 @@ def perform(model_params:, background_job_result:, signed_ids: {}, globus_ids: {
...
-      Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority) if start_workflow
-    end
-
-    StageGlobus.stage(globus_ids, druid) do
-      Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority) if start_workflow
+    if signed_ids.any?
+      StageBlobs.stage(signed_ids, druid) do
+        Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority) if start_workflow
+      end
+    elsif globus_ids.any?
```
Comment (justinlittman) [inline app/jobs/ingest_job.rb:55]: It should be possible to have both signed_ids and globus_ids.

Comment (justinlittman) [reply app/jobs/ingest_job.rb:55]: I think you probably want something like:
```
StageBlobs.stage(signed_ids, druid)
StageGlobus.stage(globus_ids, druid)
Workflow.create_unless_exists(druid, 'accessionWF', version: 1, priority: priority) if start_workflow && (signed_ids.any? || globus_ids.any?)
```

Comment (justinlittman) [reply app/jobs/ingest_job.rb:55]: I can't really figure out the purpose of the block in StageBlobs and StageGlobus. I'd recommend getting rid of the block support in these.

Comment (edsu) [reply app/jobs/ingest_job.rb:55]: @justinlittman I thought the same at first but then noticed that the AccessionWF would only be kicked off if the Staging succeeded?

Comment (edsu) [reply app/jobs/ingest_job.rb]: Ok, I've removed the block from both StageGlobus and StageBlobs and updated the tests. It does simplify the way the jobs read.

**Review by justinlittman — APPROVED** [review]

**Pull request merged**